src/Functions/seriesPeriodDetectFFT.cpp

```diff
@@ -0,0 +1,140 @@
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & /*arguments*/) const override { return std::make_shared<DataTypeFloat64>(); }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t) const override
+    {
+        ColumnPtr array_ptr = arguments[0].column;
+        const ColumnArray * array = checkAndGetColumn<ColumnArray>(array_ptr.get());
+
+        const IColumn & src_data = array->getData();
+
+        auto res = ColumnFloat64::create(1);
+        auto & res_data = res->getData();
+
+        Float64 period;
+
+        if (executeNumber<UInt8>(src_data, period) || executeNumber<UInt16>(src_data, period) || executeNumber<UInt32>(src_data, period)
+            || executeNumber<UInt64>(src_data, period) || executeNumber<Int8>(src_data, period) || executeNumber<Int16>(src_data, period)
+            || executeNumber<Int32>(src_data, period) || executeNumber<Int64>(src_data, period) || executeNumber<Float32>(src_data, period)
+            || executeNumber<Float64>(src_data, period))
+        {
+            res_data[0] = period;
+            return res;
+        }
+        else
+            throw Exception(
+                ErrorCodes::ILLEGAL_COLUMN,
+                "Illegal column {} of first argument of function {}",
+                arguments[0].column->getName(),
+                getName());
+    }
+
+    template <typename T>
+    bool executeNumber(const IColumn & src_data, Float64 & period) const
+    {
+        const ColumnVector<T> * src_data_concrete = checkAndGetColumn<ColumnVector<T>>(&src_data);
+        if (!src_data_concrete)
+            return false;
+
+        const PaddedPODArray<T> & src_vec = src_data_concrete->getData();
+
+        size_t len = src_vec.size();
+        if (len < 4)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "At least four data points are needed for function {}", getName());
+
+        std::vector<Float64> src(src_vec.begin(), src_vec.end());
+        std::vector<std::complex<double>> out((len / 2) + 1);
+
+        pocketfft::shape_t shape{len};
+
+        pocketfft::shape_t axes;
+        axes.reserve(shape.size());
+        for (size_t i = 0; i < shape.size(); ++i)
+            axes.push_back(i);
+
+        pocketfft::stride_t stride_src{sizeof(double)};
+        pocketfft::stride_t stride_out{sizeof(std::complex<double>)};
+
+        pocketfft::r2c(shape, stride_src, stride_out, axes, pocketfft::FORWARD, src.data(), out.data(), static_cast<double>(1));
+
+        size_t specLen = (len - 1) / 2; //removing the nyquist element when len is even
+
+        double maxMag = 0;
+        size_t idx = 1;
+        for (size_t i = 1; i < specLen; ++i)
+        {
+            double magnitude = sqrt(out[i].real() * out[i].real() + out[i].imag() * out[i].imag());
```

Comment:
`sqrt` can be omitted because `maxMag` is used only to get the frequency index. The square of the magnitude (`maxMagnituteSquare`?) works just fine for the purpose.